### DISCUSSION TITLE
How to customize SELECT

### DISCUSSION THREAD

**Discussion by li-up-up-up** [diesel-rs/diesel#4226] [Q&A]
```sql
SELECT ta.*, (tf1.followed_author_id NOTNULL) as is_followed, (tf.author_id NOTNULL) as is_follow
FROM tb_author ta
         LEFT JOIN tb_follow tf on ta.id = tf.author_id
         LEFT JOIN tb_follow tf1 on ta.id = tf1.followed_author_id
WHERE tf.followed_author_id = 1;
```
```rust
#[derive(Serialize, Queryable)]
pub struct FollowAuthor {
    #[serde(flatten)]
    #[diesel(embed)]
    pub author: AuthorBase,
    pub is_follow: bool,
    pub is_followed: bool,
}
pub fn find_author_follow_list(req: &FollowListReq) -> anyhow::Result<PageHelper<FollowAuthor>> {
    let order_by = (tb_author::level.desc(), tb_follow::created_at.desc());
    let tb1 = diesel::alias!(tb_follow as tb1);
    let mut search = tb_author::table
        .inner_join(tb_follow::table.on(tb_author::id.eq(tb_follow::followed_author_id)))
        .inner_join(tb1.on(tb_author::id.eq(tb1.field(tb_follow::author_id))))
        .select(FollowAuthor::as_select())
        .into_boxed();
    search = match req.typ {
        FollowTyp::Follow => search.filter(tb_follow::author_id.eq(req.author_id)),
        FollowTyp::Followed => {
            search.filter(tb1.fields(tb_follow::followed_author_id).eq(req.author_id))
        }
        FollowTyp::Inter => search
            .filter(tb_follow::followed_author_id.eq(req.author_id))
            .filter(tb1.fields(tb_follow::followed_author_id).eq(req.author_id)),
    };
    Ok(search
        .paginate(req.page, req.page_size)
        .load_and_count_pages(&mut get_db_connection()?)?)
}

```

**Comment by weiznich** [comment]
It's not clear what you want to ask. Please always include the following details in your questions if you want help:

* Your diesel version
* Your actual question
* A complete minimal reproducible example of your problem (something where someone can just perform a `cargo check/run` and see the same problem as you
* The complete error message as emitted by rustc if there is an compilation error
* A description what you've already tried so far

**Reply by li-up-up-up** [reply]
```sql
SELECT ta.*, (tf1.followed_author_id NOTNULL) as is_followed, (tf.author_id NOTNULL) as is_follow
```
TO struct, I don't know how to achieve.

**Reply by weiznich** [reply]
I'm sorry, but it is really not possible to help you. Please try to write a complete question. If you have problems with the English language, please use a online translation tool. It's still ways better to have a translated question than nothing.

**Reply by li-up-up-up** [reply]
I just want to customize the SELECT,, 
```SQL
SELECT ta.*, (tf1.followed_author_id NOTNULL) as is_followed, (tf.author_id NOTNULL) as is_follow
FROM tb_author ta
         LEFT JOIN tb_follow tf on ta.id = tf.author_id
         LEFT JOIN tb_follow tf1 on ta.id = tf1.followed_author_id
WHERE tf.followed_author_id = 1;
```
```rust
#[derive(Serialize, Queryable)]
pub struct FollowAuthor {
    #[serde(flatten)]
    #[diesel(embed)]
    pub author: AuthorBase,
    pub is_follow: bool,
    pub is_followed: bool,
}
```
and the two bool types are added to the FollowAuthor struct of the SELECT map.
But I couldn't find the documentation.

**Reply by li-up-up-up** [reply]
How to map SQL SELECT to rust structs

**Reply by li-up-up-up** [reply]
The corresponding fields are is_follow and is_followed. The corresponding struct is FollowAuthor. The corresponding function is FollowAuthor::as_select(). The following code is an error, but it is a function I want to express. I want to achieve the following code to express the ability, how should I do? The diesel version is newest.
```rust
#[derive(Serialize, Queryable)]
pub struct FollowAuthor {
    #[serde(flatten)]
    #[diesel(embed)]
    pub author: AuthorBase,
    #[diesel(sql = "(tf1.followed_author_id NOTNULL) as is_followed")]
    pub is_follow: bool,
    #[diesel(sql = "(tf.author_id NOTNULL) as is_follow")]
    pub is_followed: bool,
}
```

**Reply by weiznich** [reply]
Let me just point again to my previous request for more details: https://github.com/diesel-rs/diesel/discussions/4226#discussioncomment-10528395 It is really not possible to help you without having all these details. This especially includes the implementation of the `paginate()` and `load_and_count_pages` function + their return types.

**Reply by li-up-up-up** [reply]
This is not important! I just want to know how to customize SELECT!!!!!!

**Reply by weiznich** [reply]
Nobody forces you to provide these details if you feel that they are not important, it's just that I personally cannot help you without these information. Anyway that's my last answer here. Maybe someone else can help you, I cannot do that without the information.

**Reply by li-up-up-up** [reply]
Could you give me a document? I can modify all my code. I just want to implement a custom SELECT.
